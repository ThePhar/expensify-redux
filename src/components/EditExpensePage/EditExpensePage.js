import React from "react";

const EditExpensePage = (props) => (
    <div>Editing the expense of id of {props.match.params.id}</div>
);

export default EditExpensePage;