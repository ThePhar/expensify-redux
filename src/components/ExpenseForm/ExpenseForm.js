import moment               from 'moment';
import React                from 'react';
import { SingleDatePicker } from 'react-dates';

export default class ExpenseForm extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            description: props.expense ? props.expense.description : '',
            note: props.expense ? props.expense.note : '',
            amount: props.expense ? (props.expense.amount / 100).toString() : '',
            createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
            calendarFocused: false,
            error: ''
        };
    }

    onDescriptionChange = (e) => {
        const description = e.target.value;
        this.setState(() => ({ description }));
    };

    onNoteChange = (e) => {
        const note = e.target.value;
        this.setState(() => ({ note }));
    };

    onAmountChange = (e) => {
        const amount = e.target.value;
        if (!amount || amount.match(/^\d+(\.\d{0,2})?$/)) {
            this.setState(() => ({ amount }));
        }
    };

    onDateChange = (createdAt) => {
        if (createdAt) {
            this.setState(() => ({ createdAt }));
        }
    };

    onFocusChange = ({ focused }) => {
        this.setState(() => ({ calendarFocused: focused }));
    };

    onSubmit = (e) => {
        e.preventDefault();

        if (!this.state.description || !this.state.amount) {
            this.setState(() => ({ error: 'Please provide description and amount.' }));
        } else {
            this.setState(() => ({ error: '' }));
            this.props.onSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount) * 100,
                createdAt: this.state.createdAt.valueOf(),
                note: this.state.note
            });
        }
    };

    render() {
        return (
            <div>
                {this.state.error && <p>{this.state.error}</p>}
                <form onSubmit={this.onSubmit}>
                    <input
                        type="text"
                        placeholder="Description"
                        autoFocus
                        // required
                        value={this.state.description}
                        onChange={this.onDescriptionChange}
                    /><br />
                    <input
                        type="text"
                        placeholder="Amount"
                        value={this.state.amount}
                        onChange={this.onAmountChange}
                        // required
                    /><br />
                    <SingleDatePicker
                        date={this.state.createdAt}
                        onDateChange={this.onDateChange}
                        focused={this.state.calendarFocused}
                        onFocusChange={this.onFocusChange}
                        id="date_picker"
                        numberOfMonths={1}
                        isOutsideRange={() => false}
                    /><br />
                    <textarea
                        placeholder="Add a note for your expense."
                        value={this.state.note}
                        onChange={this.onNoteChange}
                    /><br />
                    <button>Submit</button>
                </form>
            </div>
        );
    }
}
