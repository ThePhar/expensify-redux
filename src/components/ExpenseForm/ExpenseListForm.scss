.form-layout {
    display: flex;
    align-items: center;
    justify-content: center;
    width: 100vw;

    form {
        width: 100%;
        padding: 0 1rem;
    }

    input {
        border: 1px solid #dbdbdb;
        font-weight: 200;
        font-size: 19px;
        line-height: 24px;
        color: #484848;
        padding: 11px;
        height: 48px;
        width: 100%;
        margin-bottom: 0.5rem;
    }

    textarea {
        border: 1px solid #dbdbdb;
        font-weight: 200;
        font-size: 19px;
        line-height: 24px;
        color: #484848;
        padding: 11px;
        height: calc(48px * 4);
        margin-bottom: 0.5rem;
        width: 100%;
    }

    .SingleDatePickerInput__withBorder {
        border: none;
    }
}