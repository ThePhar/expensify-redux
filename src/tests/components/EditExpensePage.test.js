import { shallow }         from 'enzyme';
import React               from 'react';
import { EditExpensePage } from '../../components/EditExpensePage/EditExpensePage';
import ExpenseForm         from '../../components/ExpenseForm/ExpenseForm';
import expenses            from '../fixtures/expenses';

let editExpense, startRemoveExpense, history, wrapper;
beforeEach(() => {
    editExpense = jest.fn();
    startRemoveExpense = jest.fn();
    history = { push: jest.fn() };
    wrapper = shallow(<EditExpensePage
        editExpense={editExpense}
        startRemoveExpense={startRemoveExpense}
        history={history}
        expense={expenses[0]}
    />);
});

test('Should render EditExpensePage correctly.', () => {
    expect(wrapper).toMatchSnapshot();
});

test('Should handle onEdit', () => {
    wrapper.find(ExpenseForm).prop('onSubmit')(expenses[0]);
    expect(history.push).toHaveBeenLastCalledWith('/');
    expect(editExpense).toHaveBeenLastCalledWith(expenses[0].id, expenses[0]);
});

test('Should handle onRemove', () => {
    wrapper.find('button').simulate('click');
    expect(history.push).toHaveBeenLastCalledWith('/');
    expect(startRemoveExpense).toHaveBeenLastCalledWith(expenses[0].id);
});