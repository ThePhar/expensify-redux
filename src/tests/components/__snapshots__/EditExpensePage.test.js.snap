// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render EditExpensePage correctly. 1`] = `
<div>
  <h1
    className="ma4"
  >
    View/Edit An Expense
  </h1>
  <ExpenseForm
    expense={
      Object {
        "amount": 195,
        "createdAt": 0,
        "description": "Gum",
        "id": "1",
        "note": "",
      }
    }
    onSubmit={[Function]}
  />
  <button
    className="button remove"
    onClick={[Function]}
  >
    Remove
  </button>
</div>
`;
